### ISSUE TITLE
Exception when using custom inline processor in normal table cell

### ISSUE THREAD

**Issue by ygra** [asciidoctor/asciidoctorj#448]
Given the sample InlineMacroProcessor (for 1.6.0) and the following document:

```asciidoc
= InlineMacroProcessor Test Document
:repo: asciidoctor/asciidoctorj-groovy-dsl

|===
|You might want to take a look at the issue issue:333[repo=asciidoctor/asciidoctorj] and issue:2[].
|===
```

will cause the following exception:

```
Exception in thread "main" java.lang.ClassCastException: org.asciidoctor.ast.impl.CellImpl cannot be cast to org.asciidoctor.ast.StructuralNode
	at org.asciidoctor.extension.processorproxies.InlineMacroProcessorProxy.process(InlineMacroProcessorProxy.java:117)
	at org.asciidoctor.extension.processorproxies.InlineMacroProcessorProxy$INVOKER$i$3$0$process.call(InlineMacroProcessorProxy$INVOKER$i$3$0$process.gen)
	at org.jruby.internal.runtime.methods.JavaMethod$JavaMethodThree.call(JavaMethod.java:1170)
	at org.jruby.internal.runtime.methods.JavaMethod$JavaMethodN.call(JavaMethod.java:729)
	at org.jruby.RubyMethod.call(RubyMethod.java:127)
	at org.jruby.RubyMethod$INVOKER$i$call.call(RubyMethod$INVOKER$i$call.gen)
	at org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:273)
	at org.jruby.runtime.callsite.CachingCallSite.callBlock(CachingCallSite.java:79)
	at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:83)
	at org.jruby.ir.instructions.CallBase.interpret(CallBase.java:419)
	at org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:322)
        ... (skipped approx. 8978923797238947 lines here)
```

This does not happen if the table cell has `a` as style to allow for arbitrary content. Presumably because the inline content is then automatically embedded in a paragraph.

**Comment by robertpanzer** [comment]
@mojavelinux In the Ruby world a Cell is not a block (StructuralNode).
Therefore we made the Cell in AsciidoctorJ a simple node (ContentNode).
Do you have an idea how to get around this? 
Should we change the signature of `MacroProcessor.process()` from

```java
Object process(StructuralNode parent, String target, Map<String, Object> attributes)
```

to

```java
Object process(ContentNode parent, String target, Map<String, Object> attributes)
```

Does that make sense?

**Comment by mojavelinux** [comment]
I think we'll have to switch the parent to a ContentNode. The parent type here is meant to be very loose since there are so many different combinations of relationships in the tree. It's up to the extension to check the parent type and take appropriate action.

**Comment by robertpanzer** [comment]
But I think it is safe to assume that the parent for a BlockMacro is a StructuralNode, isn't it?
So I would only change the signature of the process method for InlineMacroProcessor.process.

**Comment by ygra** [comment]
@robertpanzer How would a table like

```asciidoc
|===
a|image::foo.png[]
|===
```

be represented? With the `asciidoc` cell style would contents be wrapped in a StructuralNode or would the cell be the direct parent?

**Comment by robertpanzer** [comment]
With the asciidoc cell style the parent is a StructuralNode(Block).

**Comment by mojavelinux** [comment]
Yes, for now we could limit the change to the InlineMacroProcessor. We want to pick the most generic parent we can.

Keep in mind that the InlineMacroProcessor is the most likely to change as we move towards Asciidoctor 2. That's because it currently works as a streaming processor instead of an AST transformer (it returns a string instead of an AST node). So we can be more loose with the API on InlineMacroProcessor.

**Comment by mojavelinux** [comment]
> With the asciidoc cell style the parent is a StructuralNode(Block).

Technically, the parent is a Document.

**Comment by robertpanzer** [comment]
Fixed in 532f50c.
